blobfuse/src/BlockBlobBfsClient.cpp

```diff
@@ -0,0 +1,984 @@
+        }
+        return std::make_shared<blob_client_wrapper>(blobClient);
+    }
+    catch (const std::exception &ex)
+    {
+        syslog(LOG_ERR, "Failed to create blob client.  ex.what() = %s. Please check your account name and ", ex.what());
+        errno = blobfuse_constants::unknown_error;
+        return std::make_shared<blob_client_wrapper>(false);
+    }
+}
+std::shared_ptr<blob_client_wrapper> BlockBlobBfsClient::authenticate_blob_spn()
+{
+    syslog(LOG_DEBUG, "Authenticating using MSI");
+    try
+    {
+        //1. get oauth token
+        std::function<OAuthToken(std::shared_ptr<CurlEasyClient>)> SPNCallback = SetUpSPNCallback(
+            configurations.spnTenantId,
+            configurations.spnClientId,
+            configurations.spnClientSecret,
+            configurations.aadEndpoint);
+
+        std::shared_ptr<OAuthTokenCredentialManager> tokenManager = GetTokenManagerInstance(SPNCallback);
+
+        if (!tokenManager->is_valid_connection())
+        {
+            // todo: isolate definitions of errno's for this function so we can output something meaningful.
+            errno = 1;
+            return std::make_shared<blob_client_wrapper>(false);
+        }
+
+        //2. try to make blob client wrapper using oauth token
+        // We should pass the token obtained earlier to this token_credentials
+        std::shared_ptr<storage_credential> cred = std::make_shared<token_credential>("");
+        cred->set_token_callback(&GetTokenCallback);
+
+        std::shared_ptr<storage_account> account = std::make_shared<storage_account>(
+            configurations.accountName,
+            cred,
+            true, //use_https must be true to use oauth
+            configurations.blobEndpoint);
+        std::shared_ptr<blob_client> blobClient =
+            std::make_shared<blob_client>(account, max_concurrency_oauth);
+        errno = 0;
+        if (configurations.useAttrCache)
+        {
+            return std::make_shared<blob_client_attr_cache_wrapper>(std::make_shared<blob_client_wrapper>(blobClient));
+        }
+        return std::make_shared<blob_client_wrapper>(blobClient);
+    }
+    catch (const std::exception &ex)
+    {
+        syslog(LOG_ERR, "Failed to create blob client.  ex.what() = %s. Please check your account name and ", ex.what());
+        errno = blobfuse_constants::unknown_error;
+        return std::make_shared<blob_client_wrapper>(false);
+    }
+}
+
+///<summary>
+/// Uploads contents of a file to a block blob to the Storage service
+///</summary>
+///TODO: params
+///<returns>none</returns>
+void BlockBlobBfsClient::UploadFromFile(const std::string sourcePath, METADATA &metadata)
+{
+    std::string blobName = sourcePath.substr(configurations.tmpPath.size() + 6 /* there are six characters in "/root/" */);
+    InvalidateCachedProperty(blobName);
+    m_blob_client->upload_file_to_blob(sourcePath, configurations.containerName, blobName, metadata);
```

Comment:
there is one bug here. We are not able to create empty files. for eg: touch ~/mntblobfuse/mkdir11/tobedeted.txt shows successful but does not create the file in storage only creates it locally.